clang/test/CheckedC/parsing/valid-where-clause.c

```diff
@@ -30,3 +31,33 @@ void valid_cases(_Nt_array_ptr<char> p, _Nt_array_ptr<char> q,
   _Where (((((a == 0)))));
   _Where (a == 0) _And ((a == 0)) _And (((a == 0)));
 }
+
+int f();
+// Test where clauses on variable declarations inside a function.
+void valid_cases_decl(_Nt_array_ptr<char> p, _Nt_array_ptr<char> q) {
+  int a _Where a == 0 _And a != 0 _And p : bounds(p, p + a);
+  int b = 0 _Where b != 0 _And p : count(b);
+  int c = f() _Where p : bounds(p, p + c) _And c < 0;
+  int d _Where p : bounds(p, p + d) _And q : count(d) _And d == 0;
+  int e, f _Where e == 0 _And f == 0;
+  int g _Where g == 0, h, i, j _Where j < 0;
+  int k _Where k != 0 _And p : count(k), m, n, o _Where q : bounds(q, q + 1), r, s;
+  int arr1[3] = {1, 2, 3} _Where 0 < 1, arr2[2] = {1, 2} _Where 1 > 0;
+  _Nt_array_ptr<char> p1 : count(0) = "" _Where p1 : bounds(p1, p1 + 1);
+}
+
+// Test where clauses on variable declarations outside a function.
+_Nt_array_ptr<char> p;
+_Nt_array_ptr<char> q;
+int a _Where a == 0 _And a != 0 _And p : bounds(p, p + a);
+int b = 0 _Where b != 0 _And p : count(b);
+int c _Where p : bounds(p, p + c) _And q : count(c) _And c == 0;
+int d, e _Where e == 0 _And d != 0;
+int g _Where g == 0, h, i, j _Where j < 0;
+int k _Where k != 0 _And p : count(k), m, n, o _Where q : bounds(q, q + 1), r, s;
+int arr1[3] = {1, 2, 3} _Where 0 < 1, arr2[2] = {1, 2} _Where 1 > 0;
+_Nt_array_ptr<char> p1 : count(0) = "" _Where p1 : bounds(p1, p1 + 1);
+
+// Test where clauses on function parameters.
+void f1(int a _Where a < 0, int b, int c _Where c < 0, int *d : itype(_Ptr<int>) _Where d == 0) {}
+void f2(int a _Where a < 0, int b, int c _Where c < 0, int *d : itype(_Ptr<int>) _Where d == 0);
```

Comment:
We may need a few more realistic test cases with bounds-safe interfaces with where clauses that refer to parameters that appear later in the function definition/declaration.